ADBDEV-4728: Fix the check for address availability when setting up UDP sockets in IPv6 disabled environments

bandetto opened the pull request:
Fix the check for address availability when setting up UDP sockets in IPv6
disabled environments

This patch brings changes introduced in 7f3c91f (which backported 70306db,
1ee34a4) and fixes them by correctly handling IPv6 addresses.

In these patches, a fix was created for `UDPIFC` that reworked
`SendDummyPacket()` by removing the unnecessary call to `getaddrinfo()` to avoid
using `malloc()` in a signal handler, making `setupUDPListeningSocket` cache the
address of the socket used for the `UDPIFC` interconnect when it was first
created.

In IPv6 disabled environments, `bind()` with address of an IPv6 wildcard will
result in binding to both IPv6 and IPv4 (since IPv6 is disabled, only IPv4 will
be bound) on all interfaces, and it won’t fail even if IPv6 is disabled. This
mistakenly passed the check and cached address with IPv6 address family. In IPv6
disabled environments, any `connect()` or `sendto()` that reuse this address
would fail. This fail occured when we wanted to send a dummy packet with
`SendDummyPacket()` using the cached address.

Add `AI_ADDRCONFIG` flag to `getaddrinfo()` hints to force it to check for
availability of the address families. This flag will prevent `getaddrinfo()`
from returning addresses that belong to the IPv6 family when IPv6 is not
available on the host machine.

This problem is present only when `INTERCONNECT_ADDRESS_TYPE_WILDCARD` is used
as the interconnect address type. It does not affect 7.X, because the default
setting for interconnect address type was changed to
`INTERCONNECT_ADDRESS_TYPE_UNICAST`.

This is a rework of cherry-picked commit 7f3c91f.

Co-authored-by: Marbin Tan <tmarbin@vmware.com>

Chibin committed:
Remove `getaddrinfo` in `SendDummyPacket()` to address malloc deadlock

`SendDummyPacket` eventually calls `getaddrinfo` (which is a reentrant),
however, `getaddrinfo` is not an async-signal-safe function.
`getaddrinfo` internally calls `malloc`, which is strongly advised to
not do within a signal handler as it may cause deadlocks.
Cache the accepted socket information for the listener, so that it can
be reused in `SendDummyPacket()`.

The purpose of `SendDummyPacket` is to exit more quickly; it
circumvents the polling that happens, which eventually times out after 250ms.

Without `SendDummyPacket()`, there will be multiple test failures since
some tests expects the backend connection to terminate almost
immediately.

To view all the async-signal-safe functions, please view
the signal-safety(7) — Linux manual page.

====================

Accommodate for AF_INET6 and AF_INET when doing a motion layer IPC teardown

Previously on commit 70306db18e2, we removed pg_getaddrinfo_all for
signal handlers. However, in doing so, the capability of supporting both
AF_INET6 and AF_INET was lost; this responsibility must now be handled
by us. The commit mentioned above fixed the issue for AF_INET (IPv4),
but not for AF_INET6 (IPv6).

This commit also addresses the situation for both AF_INET and AF_INET6.

====================

This commit backports: 70306db18e2bfac4969bf327f5ada45bf33f57a1 and
1ee34a424548e759e82c93cf9b4fe1411465e0e9.
Conflicts resolved:
- setupUDPListeningSocket is not one to one with 7X, so resolve conflict
  on the location where `listenerSockaddr` is `memcpy`ed.
- resolve conflicts due to difference of from the modernized version and
  refactor from 50748b7

Reviewed-by: Soumyadeep Chakraborty <soumyadeep2007@gmail.com>

RekGRpth commented:
[interesting approach](https://github.com/greenplum-db/libusual/commit/b04f97fad450b9b93ead9ec5080bf4efa1f3f0fe#diff-d28c714b8d8d5bbab33f42aed59e88730c0e9dd4cf194fa1a9757e01aea508c6R87-R114)

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62435

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/993108

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/993109

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/993100

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/993103

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1000844

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1000845

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1000837

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62610

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1001885

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1001886

RekGRpth commented on src/backend/cdb/motion/test/Makefile:
```diff
@@ -0,0 +1,9 @@
+top_builddir=../../../../..
+include $(top_builddir)/src/Makefile.global
+
+TARGETS=cdbsenddummypacket
```
How do these tests run in CI?

bandetto replied:
They run in the "Build for xxx" job. But there is a caveat: all our test machines have IPv6 enabled, so [the IPv4 branch](https://github.com/arenadata/gpdb/pull/679/files#diff-b9a680f7851e2128e23bfbde09f08cc0a5c3ab94858cdfb0ec283ab9f6668313R226-R233) inside the `test_send_dummy_packet_default_wildcard()` will not be covered by our CI.

RekGRpth replied:
> They run in the "Build for xxx" job.

OK, I see

RekGRpth replied:
> all our test machines have IPv6 enabled, so [the IPv4 branch](https://github.com/arenadata/gpdb/pull/679/files#diff-b9a680f7851e2128e23bfbde09f08cc0a5c3ab94858cdfb0ec283ab9f6668313R226-R233) inside the `test_send_dummy_packet_default_wildcard()` will not be covered by our CI.

I tested it manually

andr-sokolov commented:
@bandetto , replace revert-revert with cherry-pick. Combine 2 last commits into one

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1012657

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62915

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013850

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013851

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013840

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013841

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1013842

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63063

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1022918

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1022916

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1022919

KnightMurloc commented on src/backend/cdb/motion/test/cdbsenddummypacket_test.c:
```diff
@@ -148,10 +150,95 @@ create_sender_socket(sa_family_t af)
+{
+	const struct sockaddr_in6 test_addr = {
+		.sin6_family = AF_INET6,
+		.sin6_addr = in6addr_loopback
```
why are we checking on the loopback interface? The test passes without changes in ic_udpifc.c if ipv6 is enabled only on the loopback interface, is this the expected behavior?

bandetto replied:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/freeaddrinfo.html

> If the AI_PASSIVE flag is specified, the returned address information shall be suitable for use in binding a socket for accepting incoming connections for the specified service. In this case, if the nodename argument is null, then the IP address portion of the socket address structure shall be set to INADDR_ANY for an IPv4 address or IN6ADDR_ANY_INIT for an IPv6 address. If the AI_PASSIVE flag is not specified, the returned address information shall be suitable for a call to [connect()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/connect.html) (for a connection-mode protocol) or for a call to [connect()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/connect.html), [sendto()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sendto.html), or [sendmsg()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sendmsg.html) (for a connectionless protocol). In this case, if the nodename argument is null, then the IP address portion of the socket address structure shall be set to the loopback address. The AI_PASSIVE flag shall be ignored if the nodename argument is not null.

We pass `AI_PASSIVE`, if node is `NULL`:
https://github.com/arenadata/gpdb/blob/cbdcfc62d9b6ef1462c609d56dafad8d133ccb47/src/backend/cdb/motion/ic_udpifc.c#L1226-L1227

That means we are calling `getaddrinfo()` with `in6addr_any`. `man` says:
> If hints.ai_flags includes the AI_ADDRCONFIG flag, then IPv6 addresses are returned only if the local system has at least one IPv6 address configured.  The loopback address is not considered for this case as valid as a configured address.

So you are right, and this will be a edge case when IPv6 is only enabled on a loopback device. We can check for IPv6 support excluding loopback devices with `getifaddrs()` instead.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64361

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1064935

KnightMurloc commented:
Why does the problem only affect SendDummyPacket and does not appear, for example, with Redistribute Motion?

bandetto commented:
Because DBMS remaps them to IPv4. For example, IPv4 wildcard remapped to IPv6 looks like `0:0:0:0:0:ffff:0:0`:
https://github.com/arenadata/gpdb/blob/cb99eab6234c784b114f4bfe0cefed03c06e379d/src/backend/cdb/motion/ic_udpifc.c#L2846-L2852

From `man ipv6`:
> IPv4  connections  can be handled with the v6 API by using the v4-mapped-on-v6 address type; thus a program needs to support only this API type to support both protocols. This is handled transparently by the address handling functions in the C library.

As I have tested, connectionless (UDP) IPv6 sockets that use IPv4-to-IPv6 mapped addresses work fine even if IPv6 is disabled.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64485

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1068638

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1068639

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1068636

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64573

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1071141

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1071142

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64638

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1073231

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1073232

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65448

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1107768

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1107769

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1107763

andr-sokolov commented on src/backend/cdb/motion/test/cdbsenddummypacket_test.c:
```diff
@@ -148,10 +151,97 @@ create_sender_socket(sa_family_t af)
+		fail();
+	}
+
+	for (ifaddr = ifaddr_first; ifaddr != NULL; ifaddr = ifaddr->ifa_next)
```
Variables should be as local as possible. Declare `ifaddr` in the loop.

andr-sokolov commented on src/backend/cdb/motion/test/cdbsenddummypacket_test.c:
```diff
@@ -148,10 +151,97 @@ create_sender_socket(sa_family_t af)
+
+	for (ifaddr = ifaddr_first; ifaddr != NULL; ifaddr = ifaddr->ifa_next)
+	{
+		if (ifaddr->ifa_addr == NULL)
```
Combine all the three condition into single one to remove `continue`-s. It will be condition of IPv6 presence

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65579

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1112429

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1112428

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1112430

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1112423

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1115433

bandetto committed:
Fix the check for address availability when setting up UDP sockets in IPv6
disabled environments

Add `AI_ADDRCONFIG` flag to `getaddrinfo()` hints to force it to check for
availability of the address families. This flag will prevent `getaddrinfo()`
from returning addresses that belong to the IPv6 family when IPv6 is not
available on the host machine.

This problem is present only when `INTERCONNECT_ADDRESS_TYPE_WILDCARD` is used
as the interconnect address type. It does not affect 7.X, because the default
setting for interconnect address type was changed to
`INTERCONNECT_ADDRESS_TYPE_UNICAST`.

bandetto committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4728

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/65687

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1116537

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1116538

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1116539

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1116536
